### PR TITLE
Filter duplicates at backup_database class

### LINKED ISSUE #2442: SQL Historian Locks up when handling duplicate data with Postgres
### Description of Issue
If data is received by the SQLHistorian with the same topic and timestamp in the same batch, inserting to the DB will fail on "CardinalityViolation: ON CONFLICT DO UPDATE command cannot affect row a second time"

### Affected Version
All

#### Expected
Data should be processed to prevent this error.

#### Actual
Historian fails to insert with below error.

### Steps to Reproduce
Disconnect Historian from DB so that it begins to cache in the local cache
Send multiple identical topic,ts pairs
Reconnect to to DB

### Additional Details
Aug 20 12:55:18 ip-10-1-2-171 python[15055]: 2020-08-20 12:55:16,199 (sqlhistorianagent-3.7.0 9516) volttron.platform.agent.base_historian ERROR: An unhandled exception occurred while publishing.
Aug 20 12:55:18 ip-10-1-2-171 python[15055]: Traceback (most recent call last):
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:   File "/var/lib/volttron/volttron/platform/agent/base_historian.py", line 1113, in _do_process_loop
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:     self.publish_to_historian(to_publish_list)
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:   File "/var/lib/volttron/volttron/utils/docs.py", line 47, in f
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:     return self.mthd(obj, *args, **kwargs)
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:   File "/home/volttron/.volttron/agents/54aa5cfd-09ec-437e-9253-96bd66750e48/sqlhistorianagent-3.7.0/sqlhistorian/historian.py", line 217, in publish_to_historian
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:     published += 1
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:   File "/usr/lib/python2.7/contextlib.py", line 24, in __exit__
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:     self.gen.next()
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:   File "/var/lib/volttron/volttron/platform/dbutils/postgresqlfuncts.py", line 102, in bulk_insert
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:     self.execute_stmt(query)
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:   File "/var/lib/volttron/volttron/platform/dbutils/basedb.py", line 468, in execute_stmt
Aug 20 12:55:18 ip-10-1-2-171 python[15055]:     cursor.execute(stmt, args)
Aug 20 12:55:18 ip-10-1-2-171 python[15055]: CardinalityViolation: ON CONFLICT DO UPDATE command cannot affect row a second time
Aug 20 12:55:18 ip-10-1-2-171 python[15055]: HINT:  Ensure that no rows proposed for insertion within the same command have duplicate constrained values.

### PR DESCRIPTION
**Pull request by bonicim**
# Description

Attempts to fix data duplication error when an Historian inserts duplicate data in one transaction. In response to @craig8's [suggestion](https://github.com/VOLTTRON/volttron/pull/2443#issuecomment-679308718) of fixing the problem upstream in the BackupDatabase class, I have taken @acedrew's solution in #2443 and applied it to the backup_database class. I have refactored the method `get_outstanding_to_publish` to guarantee that the output will always be a unique list of data, (assuming the idea of "sameness" or "duplicateness" is defined as having the same topic and timestamp). Marking this PR as WIP to get feedback. 

Fixes #2442  
Related to #2443 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

I have created a unit-ish test that initiliazes a real sqlite database with duplicate data  and then calls the `get_outstanding_to_publish` which is the method that sends data to the concrete historians. 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by acedrew** [comment]
This looks great, I couldn't remember if the implementation of set was O(1), but I knew dict was, that's why I used it in the heat of the moment, it looks like at this point, set has equal time performance, and far superior memory utilization to dict. 

Thanks for addressing this so quickly!

**Comment by craig8** [comment]
@acedrew are you in a position to test this with your implementation, without your fix.  I am going to try it on my end and then we can move it into the develop proper.

**Comment by bonicim** [comment]
I think there needs to be one more change to another method to make this complete. @craig8 mentioned it yesterday in that the backup_db needs to remove all duplicates once the publication is successful. I believe the method [`remove_successfully_published`](https://github.com/VOLTTRON/volttron/blob/master/volttron/platform/agent/base_historian.py#L1514_) needs to be modified  to delete all records that have the same topic and timestamp instead of a an `_id` field. Right now, if this is not changed, then the duplicates will remain in the cache and thus subject to being published again in the future.

**Comment by bonicim** [comment]
Added one more change on backup_database in which all duplicates are removed from the cache after successful publication by historian.

File: volttron/platform/agent/base_historian.py [hunk 1]
```diff
@@ -1540,19 +1540,25 @@ def remove_successfully_published(self, successful_publishes,
...
+            self._record_count -= c.rowcount
 
+        _log.debug(f"Count of rows in cache after deletion: {self._record_count}")
         self._connection.commit()
 
     def get_outstanding_to_publish(self, size_limit):
         """
-        Retrieve up to `size_limit` records from the cache.
+        Retrieve up to `size_limit` records from the cache. Guarantees a unique list of records,
+        where unique is defined as topic + timestamp.
```
Comment (craig8) [inline volttron/platform/agent/base_historian.py:1561]: topic + timestamp is assuming concatenation...use tuple syntax (topic, timestamp) .

File: volttrontesting/platform/dbutils/test_backup_database.py [hunk 2]
```diff
@@ -57,8 +59,69 @@ def test_backup_new_data_should_filter_duplicates(backup_database):
...
+        "3|2020-05-25 08:46:00|foobar_source|2|846|{}",
+    ]
+    assert outstanding_before == get_all_data("outstanding")
+
+    successful_publishes = {("2020-06-01 12:30:59", "dupetopic")}
+    expected_outstanding_after = ["3|2020-05-25 08:46:00|foobar_source|2|846|{}"]
+
+    backup_database.remove_successfully_published(successful_publishes, 1000)
+
+    assert expected_outstanding_after == get_all_data("outstanding")
```
Comment (craig8) [inline volttrontesting/platform/dbutils/test_backup_database.py:99]: I am wondering use id(expected_outstanding_after) != id(get_all_data("outstanding")) is true...the references in list could be the same and it shouldn't be after the serialize/deserialize from the database.

File: volttrontesting/platform/dbutils/test_backup_database.py [hunk 3]
```diff
@@ -57,8 +59,69 @@ def test_backup_new_data_should_filter_duplicates(backup_database):
...
+
+
+def get_all_data(table):
+    q = f'''"SELECT * FROM {table}"'''
+    res = query_db(q)
+    return res.splitlines()
+
+
+def query_db(query):
+    output = subprocess.run(
```
Comment (craig8) [inline volttrontesting/platform/dbutils/test_backup_database.py:120]: I think you should be able to run this without shell = True

File: volttron/platform/agent/base_historian.py [hunk 4]
```diff
@@ -1231,9 +1231,9 @@ def report_handled(self, record):
         """
         if isinstance(record, list):
             for x in record:
-                self._successful_published.add(x['_id'])
+                self._successful_published.add((x['timestamp'], x['topic']))
```
Comment (craig8) [inline volttron/platform/agent/base_historian.py:1234]: I don't think this needed to change in the thoughts I had...I dont' think this will be good for cache as the two columns that are mentioned in the deletion is not (nor should they) be indexed.

Comment (craig8) [reply volttron/platform/agent/base_historian.py:1234]: I think that the data should be removed from the database right as a duplicate is detected rather than waiting for a record to be passed.

File: volttrontesting/platform/dbutils/test_backup_database.py (hunk 2 above)
Comment (bonicim) [reply volttrontesting/platform/dbutils/test_backup_database.py:99]: I just tried checking the identity of both both object and yes, `id(expected_outstanding_after) != id(get_all_data("outstanding")) is true as expected. Then I checked the id's inside both lists and they are not the same either. Are you suggesting to check for reference on top of value in order to ensure that both objects are not the same thing?

**Comment by acedrew** [comment]
> Added one more change on backup_database in which all duplicates are removed from the cache after successful publication by historian.

I'm not sure this is desired behavior, we want them to be published in order of reception, (assuming the outstanding table is ordered correctly) that way historian specific duplicate handling can happen in the historian.
For example, someone might want to on duplicate avg value or trigger an insert into a duplicate table etc. The problem is simply keeping those entries in separate transactions.

**Comment by craig8** [comment]
@acedrew does my comment address your concern....I think you are right

File: volttrontesting/platform/dbutils/test_backup_database.py (hunk 3 above)
Comment (bonicim) [reply volttrontesting/platform/dbutils/test_backup_database.py:120]: You're right. I just reviewed the docs on the subprocess module and realized that I should avoid `shell = True` in general.

**Comment by craig8** [comment]
Duplicates always have been last one wins within a historian context... Generally duplicates are regarded as mistakes... We need to either make duplicates immutable. How did you actually run into this issue

File: volttron/platform/agent/base_historian.py (hunk 4 above)
Comment (bonicim) [reply volttron/platform/agent/base_historian.py:1234]: I see what you mean. When I saw that the field `_successful_published` was being used to remove things from the cache, I went in the direction of removing dupes under the `remove_successfully_published` method. Will change this and remove the dupes at the point of discovery.

File: volttron/platform/agent/base_historian.py [hunk 5]
```diff
@@ -1580,7 +1575,14 @@ def get_outstanding_to_publish(self, size_limit):
             headers = {} if row[5] is None else loads(row[5])
             meta = self._meta_data[(source, topic_id)].copy()
             topic = self._backup_cache[topic_id]
+
             if (topic, timestamp) in dedup:
+                _log.debug(f"Found duplicate; removing record from cache: {row}")
```
Comment (craig8) [inline volttron/platform/agent/base_historian.py:1580]: These should be batched so that they can be deleted all through one transaction.

**Comment by acedrew** [comment]
@craig8 for the SQL Historians, the enforcement of "last one wins" is the `ON CONFLICT (topic_id, ts) DO UPDATE SET value = EXCLUDED.value;` piece of the query. In theory one could replace that with `ON CONFLICT (topic_id, ts) DO UPDATE SET value = (value + EXCLUDED.value) / 2.0;` or some other update logic. If we drop the values before sending them to the DB, it would prevent these sorts of custom update logics from working. When encountering a duplicate value, it should ignore it in the current transaction, try again in a subsequent transaction, allowing the DB to manage the duplicates.
In answer to how I ran into this, I think it's because I have timestamp Naive MQTT data flowing into an instance, then being forwarded using the forward historian. The duplicates caused the issue at the top node where it was finally being sent to the SQL DB

**Comment by craig8** [comment]
@acedrew we need to discuss this do you have some time tomorrow?

File: volttron/platform/agent/base_historian.py (hunk 5 above)
Comment (bonicim) [reply volttron/platform/agent/base_historian.py]: Good point. Will change.

**Comment by acedrew** [comment]
@craig8 have you quit slack? Happy to follow up and discuss.

**Comment by shwethanidd** [comment]
@acedrew Craig is OOO. We'll set something up when he is back tomorrow.

**Comment by bonicim** [comment]
After meeting with @craig8 and @acedrew on 9/3, it was decided that duplicates should not be removed from the cache. Rather duplicates should be handled at the database level and should return duplicates to 'publish_historian' on follow-on transactions. This PR is a first attempt at doing just that. It is currently in WIP and seeking feedback on implementation and ideas for optimization/cleanup. FYI, I will be OO for the rest of the week (9/8 - 9/11).

**Comment by bonicim** [comment]
Simplified tests and added some basic tests without duplicates. This PR is now ready for review.

**Comment by bonicim** [comment]
After talking with @craig8 over the case of size_limit being a small number, I realized that instead of handling the duplicates after I received the data from the outstanding table, I could manipulate the SQL query to both return unique results and delete unique results by using the keyword "group by" which is similar to the "select distinct" SQL query. The updated solution eliminates the need to track duplicates and simply uses SQL to do deduplication. 

I have also added a test that does multiple transactions to show the expected behavior on a small size_limit value.

**Comment by bonicim** [comment]
Added a test for sqlitehistorian unit test that tests the case in which duplicates are present in the cache. The test is still a work in progress, putting them out here for feedback.

**Comment by bonicim** [comment]
As requested by @craig8 , I wrote an [sql performance script](https://gist.github.com/bonicim/9e696e287fea7bf2ebb5bbc135eda66a) to measure the difference between using 'group by' to remove duplicates using the current sql command without 'group by'. It graphs the size against the time duration. Attached is a picture of the graph. It appears that 'group by' adds about 3 seconds to the query. Two options here: 1) handle deduplication at the sql query; pros: offload work to SQL, cons: huge performance hit of about 7x; 2) handle deduplication at the Base Historian level; pros: more control, better performance maybe; cons: code bloat due to handling deduplication, increase complexity/work of Base Historian
Figure_1

**Comment by bonicim** [comment]
Based on performance test of `group by` query and the feedback from the team, deduplication will now be handled at the BaseHistorian level instead of the cache database. While this improves performance, it does add more work and complexity to BaseHistorian; I have added and updated tests on both the BaseHistorian level and a concrete historian, using and sqlite-based historian.

File: volttron/platform/agent/base_historian.py [hunk 6]
```diff
@@ -1572,30 +1587,43 @@ def get_outstanding_to_publish(self, size_limit):
             value = loads(row[4])
             headers = {} if row[5] is None else loads(row[5])
             meta = self._meta_data[(source, topic_id)].copy()
+            topic = self._backup_cache[topic_id]
+
+            # check for duplicates before appending row to results
+            if (topic, timestamp) in unique_records:
```
Comment (schandrika) [inline volttron/platform/agent/base_historian.py:1593]: why not  use just topic_id, timestamp ?

File: volttron/platform/agent/base_historian.py [hunk 7]
```diff
@@ -1518,41 +1520,54 @@ def remove_successfully_published(self, successful_publishes,
...
-                self._record_count -= c.rowcount
-        else:
-            temp = list(successful_publishes)
-            temp.sort()
-            c.executemany('''DELETE FROM outstanding
-                            WHERE id = ?''',
-                          ((_id,) for _id in
-                           successful_publishes))
-            self._record_count -= len(temp)
+                temp = list(successful_publishes)
```
Comment (schandrika) [inline volttron/platform/agent/base_historian.py:1553]: there could be duplicate records even in the case were not all records in the batch got into db.

File: volttron/platform/agent/base_historian.py [hunk 8]
```diff
@@ -1518,41 +1520,54 @@ def remove_successfully_published(self, successful_publishes,
...
+                if not self._dupe_ids:
+                    c.execute('''DELETE FROM outstanding
+                                WHERE ROWID IN
+                                (SELECT ROWID FROM outstanding ORDER BY ts LIMIT ?)''', (submit_size,))
+                    if self._record_count < c.rowcount:
+                        self._record_count = 0
+                    else:
+                        self._record_count -= c.rowcount
+                else:
+                    # this is a rare case, so adding logs to help track and debug when this case does occur
```
Comment (schandrika) [inline volttron/platform/agent/base_historian.py:1545]: we could always delete based on unique_ids unless the first query would perform better. i am not sure if that is true though.

File: services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+import pytest
+from gevent import sleep
+from datetime import timedelta
+from services.core.SQLHistorian.sqlhistorian import historian
+
+CACHE_NAME = "backup.sqlite"
+HISTORIAN_DB = "./data/historian.sqlite"
+
+
```
Comment (schandrika) [inline services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py:13]: There is a separate  test file for base historian under volttrontesting/services/historian/test_bbase_historian.py . This test case should get added to base historian tests

**Review by schandrika — CHANGES_REQUESTED** [review]

File: services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py (hunk 9 above)
Comment (craig8) [reply services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py:13]: note this is a unit test not an integration test.  it does not run the platform.

Comment (bonicim) [reply services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py:13]: To @craig8's point, the intent of this test was to verify that a concrete a historian (in this case, an historian using sqlite) was handling duplicates as expected, i.e. that the last duplicate "wins" and gets saved in the db.

File: volttron/platform/agent/base_historian.py (hunk 6 above)
Comment (bonicim) [reply volttron/platform/agent/base_historian.py:1593]: Good question. I originally used topic because I was using the solution from [@acedrew's earlier PR](https://github.com/VOLTTRON/volttron/pull/2443/files#diff-4d58a5e8d1e986cea80064a795a19e46d509c2a22d449c116651a121d3df514cR187), which doesn't use topic_id because the concrete historian doesn't have topic_id, only topic, which makes sense because topic_id comes from the cache, which only the base historian knows, but not a concrete historian. 

I've changed the code accordingly and all tests still pass. Topic_id is more concise to identify duplicates.

File: volttron/platform/agent/base_historian.py (hunk 7 above)
Comment (bonicim) [reply volttron/platform/agent/base_historian.py:1553]: Can you elaborate on this particular case? How would this happen in the historian execution flow?

File: volttron/platform/agent/base_historian.py (hunk 8 above)
Comment (bonicim) [reply volttron/platform/agent/base_historian.py:1545]: How would the query look like instead? I though I was deleting on unique_ids  using the following:

```sql
'''DELETE FROM outstanding 
  WHERE id = ?''',
((_id,) for _id in self._unique_ids)
```

Comment (schandrika) [reply volttron/platform/agent/base_historian.py:1545]: I meant you could use the above query in both if and else clause. If there are no duplicates won't unique ids simply contain all the ids the subquery (SELECT ROWID FROM outstanding ORDER BY ts LIMIT ?)''', (submit_size,)) return ?

File: services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py (hunk 9 above)
Comment (schandrika) [reply services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py:13]: This a good one to test. But since this feature is added to base historian it would be good to add a test case to base historian test cases irrespective of the concrete historian.

File: volttron/platform/agent/base_historian.py (hunk 8 above)
Comment (bonicim) [reply volttron/platform/agent/base_historian.py:1545]: I see your point. I initially did it this way to highlight and isolate the case of duplicates, maintain the supposed performance gain on executing one query in the case of no duplicates, and not delete current production code. But you are right, in both cases, the query will delete items whether it has duplicates or not.

Comment (bonicim) [reply volttron/platform/agent/base_historian.py:1545]: I did a quick and dirty performance test between execute and executemany on a DELETE query with a size of 1000 records. It appears that the difference is negligible:

`executemany`
PASSED [100%]deleted
elapsed: 0.0029566287994384766

`execute`
PASSED [100%]deleted
elapsed: 0.002218961715698242

Comment (bonicim) [reply volttron/platform/agent/base_historian.py]: Based on offline conversation with @schandrika, I've flattened the logic to be a simple if/else.

File: services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py (hunk 9 above)
Comment (bonicim) [reply services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py:13]: Agree. I added this extra test to validate our assumption that the "last one" wins in a duplicate saved to a database at the concrete historian level. The initial scope for this PR is fixing the backup_database so that duplicates aren't sent to the base_historian. I'll create a separate issue for base_historian unit tests to be tracked in a different PR.

Comment (bonicim) [reply services/core/SQLHistorian/tests/test_sqlitehistorian_unit.py:13]: Added a simple base historian agent unit test that mocks the parent agent in the latest commit, 46321f8.

**Comment by craig8** [comment]
@bonicim bump @schandrika status

**Comment by bonicim** [comment]
@schandrika Comments added.
